File: tools/session-state/src/session_state/gitfacts.py
"""gitfacts.py — collect git facts via a single porcelain-v2 call + a filesystem probe.

porcelain v2 record prefixes parsed for `dirty`:
  '1' changed, '2' renamed/copied (path is field after the score), 'u' unmerged, '?' untracked.
'# branch.head' gives the branch or the literal '(detached)'. upstream/ab headers are absent
when detached and are intentionally NOT used (ahead/behind is not tracked).
"""
from __future__ import annotations

import shutil
import subprocess
from pathlib import Path


def parse_porcelain_v2(text: str) -> dict:
    branch: str | None = None
    detached = False
    dirty: list[str] = []
    for line in text.splitlines():
        if line.startswith("# branch.head "):
            head = line[len("# branch.head "):].strip()
            if head == "(detached)":
                detached = True
                branch = None
            else:
                branch = head
        elif line.startswith("1 ") or line.startswith("2 "):
            # ordinary/renamed: path is the last tab-or-space field; renames carry "\told".
            payload = line.split("\t")[0]
            dirty.append(payload.split(" ")[-1])
        elif line.startswith("u "):
            dirty.append(line.split("\t")[0].split(" ")[-1])
        elif line.startswith("? "):
            dirty.append(line[2:].strip())
    return {"branch": branch, "detached": detached, "dirty": dirty}


def in_progress(git_dir: Path) -> str | None:
    if (git_dir / "rebase-merge").exists() or (git_dir / "rebase-apply").exists():
        return "rebase"
    if (git_dir / "MERGE_HEAD").exists():
        return "merge"
    if (git_dir / "BISECT_LOG").exists():
        return "bisect"
    return None


def _git(cwd: Path, args: list[str], git_bin: str) -> subprocess.CompletedProcess:
    return subprocess.run(
        [git_bin, "-C", str(cwd), *args],
        capture_output=True, text=True, timeout=5,
    )


def collect_git_facts(cwd: Path, dirty_cap: int = 50) -> dict | None:
    git_bin = shutil.which("git")
    if git_bin is None:
        # binary missing — visible, not silent (Spec §9 #10)
        return {"branch": None, "detached": False, "in_progress": None,
                "head": None, "dirty": [], "worktree": None, "git_unavailable": True}
    try:
        top = _git(cwd, ["rev-parse", "--show-toplevel"], git_bin)
        if top.returncode != 0:
            return None  # not a git repo
        status = _git(cwd, ["--no-optional-locks", "status", "--porcelain=v2", "--branch"], git_bin)
        parsed = parse_porcelain_v2(status.stdout)
        head_subject = None
        head = _git(cwd, ["log", "-1", "--format=%h %s"], git_bin)
        if head.returncode == 0:
            head_subject = head.stdout.strip()
        git_common = _git(cwd, ["rev-parse", "--git-dir"], git_bin)
        git_dir = Path(cwd) / git_common.stdout.strip() if git_common.returncode == 0 else Path(cwd) / ".git"
        # worktree: linked-worktree subpath if cwd's toplevel differs from the main worktree
        wt = None
        common = _git(cwd, ["rev-parse", "--path-format=absolute", "--git-common-dir"], git_bin)
        if common.returncode == 0:
            common_root = Path(common.stdout.strip()).parent
            if common_root != Path(top.stdout.strip()):
                wt = str(Path(top.stdout.strip()).name)
        return {
            "branch": parsed["branch"],
            "detached": parsed["detached"],
            "in_progress": in_progress(git_dir),
            "head": head_subject,
            "dirty": parsed["dirty"][:dirty_cap],
            "worktree": wt,
            "git_unavailable": False,
        }
    except Exception:
        return None

File: plugins/session-state/src/gitfacts.mjs
import fs from "node:fs";
import path from "node:path";
import { execFileSync } from "node:child_process";

export function parsePorcelainV2(text) {
  let branch = null, detached = false; const dirty = [];
  for (const line of text.split(/\r?\n/)) {
    if (line.startsWith("# branch.head ")) {
      const head = line.slice("# branch.head ".length).trim();
      if (head === "(detached)") { detached = true; branch = null; } else branch = head;
    } else if (line.startsWith("1 ") || line.startsWith("2 ")) {
      const left = line.split("\t")[0];
      dirty.push(left.split(" ").pop());
    } else if (line.startsWith("u ")) {
      dirty.push(line.split("\t")[0].split(" ").pop());
    } else if (line.startsWith("? ")) {
      dirty.push(line.slice(2).trim());
    }
  }
  return { branch, detached, dirty };
}

export function inProgress(gitDir) {
  const has = (n) => fs.existsSync(path.join(gitDir, n));
  if (has("rebase-merge") || has("rebase-apply")) return "rebase";
  if (has("MERGE_HEAD")) return "merge";
  if (has("BISECT_LOG")) return "bisect";
  return null;
}

function git(cwd, args) {
  return execFileSync("git", ["-C", cwd, ...args], { encoding: "utf8", timeout: 5000 });
}

export function collectGitFacts(cwd, cap = 50) {
  try { execFileSync("git", ["--version"], { timeout: 5000 }); }
  catch { return { branch: null, detached: false, in_progress: null, head: null, dirty: [], worktree: null, git_unavailable: true }; }
  let top;
  try { top = git(cwd, ["rev-parse", "--show-toplevel"]).trim(); }
  catch { return null; } // not a git repo
  if (!top) return null;
  const status = (() => { try { return git(cwd, ["--no-optional-locks", "status", "--porcelain=v2", "--branch"]); } catch { return ""; } })();
  const parsed = parsePorcelainV2(status);
  let head = null;
  try { head = git(cwd, ["log", "-1", "--format=%h %s"]).trim(); } catch {}
  let gitDir = path.join(cwd, ".git");
  try { gitDir = path.resolve(cwd, git(cwd, ["rev-parse", "--git-dir"]).trim()); } catch {}
  let worktree = null;
  try {
    const common = git(cwd, ["rev-parse", "--path-format=absolute", "--git-common-dir"]).trim();
    if (common && path.dirname(common) !== top) worktree = path.basename(top);
  } catch {}
  return {
    branch: parsed.branch, detached: parsed.detached, in_progress: inProgress(gitDir),
    head, dirty: parsed.dirty.slice(0, cap), worktree, git_unavailable: false,
  };
}
